### DISCUSSION TITLE
Error when running getting started demo: Authorization failed

### DISCUSSION THREAD

**Discussion by sagilar** [iTwin/itwinjs-core#3160] [Q&A] [answered]
Hi,  

I'm trying to set up the basic demo of the getting started with iTwin, which is the  [webviewer tutorial](https://www.itwinjs.org/learning/tutorials/develop-web-viewer/), but I have had no luck. I have followed all the steps according to the documentation, and created the Single Page Application with the IModel sample "House Sample". Then, when running the nodejs application with the `@itwin/web-viewer` template, it starts correctly, but when I try to open it and sign in to the application I get `An Error Has Occurred
The user is unauthorized. Please provide valid authentication credentials.` (see attached picture).  

![image](https://user-images.githubusercontent.com/36315612/152365053-a49ed08c-2d6c-40dd-bc33-61e883a36137.png)

I thought the implementation was going to be straightforward, as it is the _getting started_ demo, but it was not the case. I tried to follow some of the discussions on GitHub and on Bentley's iTwin forum with no luck.  

By the way, when I try to see my role privileges, I don't have any. It's weird because the project, the app, and the sample were created from this "test" account.  

Maybe I'm missing some steps, but they are definitely not specified in the documentation.  

Thanks in advance.

**Comment by aruniverse** [comment, marked as answer]
Hey,

Sorry, the tutorial you're referring to is a bit outdated now. Have you taken a look at https://developer.bentley.com/tutorials/web-application-quick-start/ ? What version of itwin/web-viewer-react are you pulling in?

Just to confirm you have a valid client with the proper scopes (`imodelaccess:read imodels:read realitydata:read`) and you do have access to the project / imodel you're trying to view?

**Reply by sagilar** [reply]
I just pulled the template with this command: npx create-react-app incubator --template @itwin/web-viewer --scripts-version @bentley/react-scripts as suggested in the guide.  

Currently, the app that I created (just for visualization) has only 2 scopes (created by default) `itwinjs imodelaccess:read`, which I added to the .env file of the application.  
I also set the Redirect URI to http://localhost:3000/signin-callback in both the app and the .env file.  
Finally, I copied and pasted the client id, imodel id, and itwin id from the browser to the .env file.  
In general terms, I just followed the guide step by step. I have double-checked a couple of times, and the result is the same; the same error.

Thanks for your help.

**Reply by aruniverse** [reply]
You can actually remove the `itwinjs` scope in favor of `imodelaccess:read`. But you will need the `imodels:read` scope to actually open the imodel.

**Reply by sagilar** [reply]
With the scope you suggested I got this: `http://localhost:3000/signin-callback?error_description=The+requested+scope+is+invalid%2C+unknown%2C+malformed%2C+or+exceeds+that+which+the+client+is+permitted+to+request.&state=ad2ccc9209124645931fbb2ea523c32e&error=invalid_scope#.` 

Btw, I just updated the .env file with the scope you suggested as I can't modify the one on the myapps webpage.

**Reply by aruniverse** [reply]
You need to add it to your client as well. 

Go to https://developer.bentley.com/apps/spa/{your client id here} 
Click on Add APIs > Digital Twin Management > iModels, should add the `imodels:read` scope.

You'll need to wait a few mins for the update to take effect server-side, but this should work.

**Reply by sagilar** [reply]
Yeah, now it's working. Thanks for your help.

**Reply by sagilar** [reply]
Now I'm getting this:

`Oops, something went wrong - please reload the page.
In case you're interested in the details:
create@http://localhost:3000/static/js/vendors~main.chunk.js:283299:34
compile@http://localhost:3000/static/js/vendors~main.chunk.js:275364:83
compileUniforms@http://localhost:3000/static/js/vendors~main.chunk.js:275612:20
compile@http://localhost:3000/static/js/vendors~main.chunk.js:275560:139
use@http://localhost:3000/static/js/vendors~main.chunk.js:275571:14
changeProgram@http://localhost:3000/static/js/vendors~main.chunk.js:276020:43
setProgram@http://localhost:3000/static/js/vendors~main.chunk.js:275957:17
execute@http://localhost:3000/static/js/vendors~main.chunk.js:262083:14
execute/<@http://localhost:3000/static/js/vendors~main.chunk.js:281343:47
using@http://localhost:3000/static/js/vendors~main.chunk.js:174326:24
using@http://localhost:3000/static/js/vendors~main.chunk.js:174319:41
execute@http://localhost:3000/static/js/vendors~main.chunk.js:281342:70
drawPass@http://localhost:3000/static/js/vendors~main.chunk.js:272906:28
renderTranslucent/<@http://localhost:3000/static/js/vendors~main.chunk.js:273231:12
execute@http://localhost:3000/static/js/vendors~main.chunk.js:263625:5
renderTranslucent@http://localhost:3000/static/js/vendors~main.chunk.js:273230:79
draw@http://localhost:3000/static/js/vendors~main.chunk.js:271999:12
paintScene@http://localhost:3000/static/js/vendors~main.chunk.js:279011:23
drawFrame@http://localhost:3000/static/js/vendors~main.chunk.js:278852:10
renderFrame@http://localhost:3000/static/js/vendors~main.chunk.js:245174:14
renderLoop@http://localhost:3000/static/js/vendors~main.chunk.js:239082:42
eventLoop@http://localhost:3000/static/js/vendors~main.chunk.js:229178:29
`

When loading the webpage with the House Sample model.

Any suggestion on this?

**Reply by pmconne** [reply]
Using the same device and browser, visit [this page](https://connect-imodeljscompatibility.bentley.com) and post a screenshot of it here.

**Reply by sagilar** [reply]
I got this:

![image](https://user-images.githubusercontent.com/36315612/152395963-445ee3cf-b5ce-4e6e-b42b-cedc9c1b441f.png)

**Reply by pmconne** [reply]
You're running in a VM? "Generic Renderer" sounds like it's probably using software rendering - that's not going to work well if at all.
Have you tried running outside a VM, and/or configuring your VM to use your graphics card?

**Reply by sagilar** [reply]
I will try it on Windows instead.

**Reply by sagilar** [reply]
It worked. Thanks for the help.
